File: js/doc/draw_react_flow.py
#!/usr/bin/env python3
"""绘制 ReAct 工作流程图"""

import matplotlib.pyplot as plt
import matplotlib.patches as patches
from matplotlib.patches import FancyBboxPatch, FancyArrowPatch
import matplotlib.lines as mlines

# 设置中文字体
plt.rcParams['font.sans-serif'] = ['Arial Unicode MS', 'SimHei', 'DejaVu Sans']
plt.rcParams['axes.unicode_minus'] = False

fig, ax = plt.subplots(1, 1, figsize=(14, 8))
ax.set_xlim(0, 14)
ax.set_ylim(0, 8)
ax.axis('off')

# 颜色定义
llm_color = '#E3F2FD'
env_color = '#F3E5F5'
thought_color = '#FFF9C4'
action_color = '#FFECB3'
pause_color = '#FFCDD2'
tool_color = '#E1BEE7'
obs_color = '#C5CAE9'
final_color = '#C8E6C9'

# LLM 层外框
llm_box = FancyBboxPatch((0.5, 4.5), 13, 3, 
                         boxstyle="round,pad=0.1", 
                         edgecolor='#1976D2', 
                         facecolor=llm_color, 
                         linewidth=2)
ax.add_patch(llm_box)
ax.text(7, 7.3, 'LLM', fontsize=14, weight='bold', ha='center')

# 执行环境外框
env_box = FancyBboxPatch((0.5, 0.5), 13, 3.5, 
                         boxstyle="round,pad=0.1", 
                         edgecolor='#7B1FA2', 
                         facecolor=env_color, 
                         linewidth=2)
ax.add_patch(env_box)
ax.text(7, 3.7, '人类/执行环境', fontsize=14, weight='bold', ha='center')

# LLM 层的节点
def draw_node(x, y, width, height, text, color):
    node = FancyBboxPatch((x-width/2, y-height/2), width, height,
                          boxstyle="round,pad=0.05",
                          edgecolor='#333',
                          facecolor=color,
                          linewidth=1.5)
    ax.add_patch(node)
    ax.text(x, y, text, fontsize=11, ha='center', va='center', weight='bold')

# Input
draw_node(1.5, 6.5, 1.5, 0.8, 'Input', thought_color)

# Thought
draw_node(3.5, 6.5, 1.5, 0.8, 'Thought', thought_color)

# Action
draw_node(6, 6.5, 1.5, 0.8, 'Action', action_color)

# Action Input
draw_node(8.5, 6.5, 1.8, 0.8, 'Action\nInput', action_color)

# PAUSE
draw_node(11.5, 6.5, 1.5, 0.8, 'PAUSE', pause_color)

# 执行环境的节点
# 工具执行
draw_node(11.5, 2.5, 1.8, 0.8, '工具执行', tool_color)

# Observation
draw_node(8.5, 2.5, 2, 0.8, 'Observation', obs_color)

# Final Answer
draw_node(3.5, 1.2, 2.2, 0.8, 'Final Answer', final_color)

# 绘制箭头
def draw_arrow(x1, y1, x2, y2, style='solid', color='#333'):
    arrow = FancyArrowPatch((x1, y1), (x2, y2),
                           arrowstyle='->', 
                           mutation_scale=20,
                           linewidth=2,
                           color=color,
                           linestyle=style)
    ax.add_patch(arrow)

# LLM 层内部流程
draw_arrow(2.25, 6.5, 2.75, 6.5)  # Input -> Thought
draw_arrow(4.25, 6.5, 5.25, 6.5)  # Thought -> Action
draw_arrow(6.75, 6.5, 7.6, 6.5)   # Action -> Action Input
draw_arrow(9.4, 6.5, 10.75, 6.5)  # Action Input -> PAUSE

# PAUSE 到工具执行
draw_arrow(11.5, 6.1, 11.5, 2.9, color='#7B1FA2')

# 工具执行到 Observation
draw_arrow(10.6, 2.5, 9.5, 2.5, color='#7B1FA2')

# Observation 回到 Thought (循环)
# 先向下，再向左，再向上
ax.plot([8.5, 8.5], [2.1, 1.5], 'k-', linewidth=2, color='#1976D2')
ax.plot([8.5, 3.5], [1.5, 1.5], 'k-', linewidth=2, color='#1976D2')
ax.plot([3.5, 3.5], [1.5, 4.0], 'k-', linewidth=2, color='#1976D2')
# 箭头
draw_arrow(3.5, 4.0, 3.5, 6.1, color='#1976D2')

# Thought 到 Final Answer (虚线，表示条件判断)
ax.plot([3.5, 3.5], [6.1, 1.6], 'k--', linewidth=2, color='#4CAF50', alpha=0.7)
draw_arrow(3.5, 1.8, 3.5, 1.6, color='#4CAF50')

# 添加文字说明
ax.text(12.5, 4.3, '①', fontsize=16, weight='bold', color='#7B1FA2')
ax.text(10, 2.8, '②', fontsize=16, weight='bold', color='#7B1FA2')
ax.text(6, 1.5, '③ 循环', fontsize=12, weight='bold', color='#1976D2')
ax.text(4.5, 4, '④ 完成', fontsize=12, weight='bold', color='#4CAF50')

plt.tight_layout()
plt.savefig('/Users/youxingzhi/ayou/Geek03/js/doc/images/react-workflow.png', 
            dpi=300, bbox_inches='tight', facecolor='white')
print("✅ 流程图已生成：/Users/youxingzhi/ayou/Geek03/js/doc/images/react-workflow.png")

File: js/doc/draw-react-flow.js
#!/usr/bin/env node
/**
 * 绘制 ReAct 工作流程图
 * 使用 Node.js Canvas API
 */

const fs = require('fs');
const path = require('path');

// 检查是否安装了 canvas
let Canvas;
try {
  Canvas = require('canvas');
} catch (e) {
  console.log('正在安装 canvas 库...');
  require('child_process').execSync('npm install canvas', { stdio: 'inherit' });
  Canvas = require('canvas');
}

const { createCanvas } = Canvas;

// 创建画布
const width = 1400;
const height = 800;
const canvas = createCanvas(width, height);
const ctx = canvas.getContext('2d');

// 填充白色背景
ctx.fillStyle = '#ffffff';
ctx.fillRect(0, 0, width, height);

// 颜色定义
const colors = {
  llmBg: '#E3F2FD',
  llmBorder: '#1976D2',
  envBg: '#F3E5F5',
  envBorder: '#7B1FA2',
  thought: '#FFF9C4',
  action: '#FFECB3',
  pause: '#FFCDD2',
  tool: '#E1BEE7',
  obs: '#C5CAE9',
  final: '#C8E6C9',
};

// 绘制圆角矩形
function roundRect(ctx, x, y, width, height, radius) {
  ctx.beginPath();
  ctx.moveTo(x + radius, y);
  ctx.lineTo(x + width - radius, y);
  ctx.quadraticCurveTo(x + width, y, x + width, y + radius);
  ctx.lineTo(x + width, y + height - radius);
  ctx.quadraticCurveTo(x + width, y + height, x + width - radius, y + height);
  ctx.lineTo(x + radius, y + height);
  ctx.quadraticCurveTo(x, y + height, x, y + height - radius);
  ctx.lineTo(x, y + radius);
  ctx.quadraticCurveTo(x, y, x + radius, y);
  ctx.closePath();
}

// 绘制外框
function drawContainer(x, y, w, h, title, bgColor, borderColor) {
  ctx.save();
  roundRect(ctx, x, y, w, h, 15);
  ctx.fillStyle = bgColor;
  ctx.fill();
  ctx.strokeStyle = borderColor;
  ctx.lineWidth = 3;
  ctx.stroke();
  
  // 绘制标题
  ctx.fillStyle = '#000000';
  ctx.font = 'bold 24px Arial';
  ctx.textAlign = 'center';
  ctx.fillText(title, x + w / 2, y + 35);
  ctx.restore();
}

// 绘制节点
function drawNode(x, y, w, h, text, bgColor) {
  ctx.save();
  roundRect(ctx, x, y, w, h, 8);
  ctx.fillStyle = bgColor;
  ctx.fill();
  ctx.strokeStyle = '#333333';
  ctx.lineWidth = 2;
  ctx.stroke();
  
  // 绘制文字
  ctx.fillStyle = '#000000';
  ctx.font = 'bold 18px Arial';
  ctx.textAlign = 'center';
  ctx.textBaseline = 'middle';
  
  // 处理多行文本
  const lines = text.split('\n');
  const lineHeight = 22;
  const startY = y + h / 2 - ((lines.length - 1) * lineHeight) / 2;
  
  lines.forEach((line, i) => {
    ctx.fillText(line, x + w / 2, startY + i * lineHeight);
  });
  
  ctx.restore();
}

// 绘制箭头
function drawArrow(x1, y1, x2, y2, color = '#333333', lineWidth = 3) {
  const headlen = 15;
  const angle = Math.atan2(y2 - y1, x2 - x1);
  
  ctx.save();
  ctx.strokeStyle = color;
  ctx.fillStyle = color;
  ctx.lineWidth = lineWidth;
  
  // 画线
  ctx.beginPath();
  ctx.moveTo(x1, y1);
  ctx.lineTo(x2, y2);
  ctx.stroke();
  
  // 画箭头
  ctx.beginPath();
  ctx.moveTo(x2, y2);
  ctx.lineTo(
    x2 - headlen * Math.cos(angle - Math.PI / 6),
    y2 - headlen * Math.sin(angle - Math.PI / 6)
  );
  ctx.lineTo(
    x2 - headlen * Math.cos(angle + Math.PI / 6),
    y2 - headlen * Math.sin(angle + Math.PI / 6)
  );
  ctx.closePath();
  ctx.fill();
  
  ctx.restore();
}

// 绘制虚线箭头
function drawDashedArrow(x1, y1, x2, y2, color = '#4CAF50') {
  const headlen = 15;
  const angle = Math.atan2(y2 - y1, x2 - x1);
  
  ctx.save();
  ctx.strokeStyle = color;
  ctx.fillStyle = color;
  ctx.lineWidth = 3;
  ctx.setLineDash([10, 5]);
  
  // 画线
  ctx.beginPath();
  ctx.moveTo(x1, y1);
  ctx.lineTo(x2, y2);
  ctx.stroke();
  
  ctx.setLineDash([]);
  
  // 画箭头
  ctx.beginPath();
  ctx.moveTo(x2, y2);
  ctx.lineTo(
    x2 - headlen * Math.cos(angle - Math.PI / 6),
    y2 - headlen * Math.sin(angle - Math.PI / 6)
  );
  ctx.lineTo(
    x2 - headlen * Math.cos(angle + Math.PI / 6),
    y2 - headlen * Math.sin(angle + Math.PI / 6)
  );
  ctx.closePath();
  ctx.fill();
  
  ctx.restore();
}

// 绘制标注
function drawLabel(x, y, text, color = '#000000', size = 20) {
  ctx.save();
  ctx.fillStyle = color;
  ctx.font = `bold ${size}px Arial`;
  ctx.textAlign = 'center';
  ctx.fillText(text, x, y);
  ctx.restore();
}

// 开始绘制
console.log('🎨 开始绘制 ReAct 流程图...');

// LLM 层外框
drawContainer(50, 450, 1300, 300, '🤖 LLM 层', colors.llmBg, colors.llmBorder);

// 执行环境外框
drawContainer(50, 50, 1300, 350, '⚙️ 人类/执行环境', colors.envBg, colors.envBorder);

// LLM 层的节点
const nodeWidth = 150;
const nodeHeight = 80;
const nodeY = 600;

drawNode(100, nodeY, nodeWidth, nodeHeight, 'Input', colors.thought);
drawNode(350, nodeY, nodeWidth, nodeHeight, 'Thought', colors.thought);
drawNode(600, nodeY, nodeWidth, nodeHeight, 'Action', colors.action);
drawNode(850, nodeY, nodeWidth, nodeHeight, 'Action\nInput', colors.action);
drawNode(1150, nodeY, nodeWidth, nodeHeight, 'PAUSE', colors.pause);

// 执行环境的节点
drawNode(1150, 250, 180, nodeHeight, '工具执行', colors.tool);
drawNode(850, 250, 200, nodeHeight, 'Observation', colors.obs);
drawNode(350, 120, 220, nodeHeight, 'Final Answer', colors.final);

// LLM 层内部流程箭头
drawArrow(250, nodeY + nodeHeight / 2, 350, nodeY + nodeHeight / 2);
drawArrow(500, nodeY + nodeHeight / 2, 600, nodeY + nodeHeight / 2);
drawArrow(750, nodeY + nodeHeight / 2, 850, nodeY + nodeHeight / 2);
drawArrow(1000, nodeY + nodeHeight / 2, 1150, nodeY + nodeHeight / 2);

// PAUSE 到工具执行
drawArrow(1225, nodeY, 1225, 330, colors.envBorder);

// 工具执行到 Observation
drawArrow(1150, 290, 1050, 290, colors.envBorder);

// Observation 回到 Thought (循环)
ctx.save();
ctx.strokeStyle = colors.llmBorder;
ctx.lineWidth = 3;
ctx.beginPath();
ctx.moveTo(850, 250);
ctx.lineTo(850, 150);
ctx.lineTo(425, 150);
ctx.lineTo(425, 400);
ctx.stroke();
ctx.restore();
drawArrow(425, 400, 425, 600, colors.llmBorder);

// Thought 到 Final Answer (虚线)
drawDashedArrow(425, nodeY, 425, 200, colors.final);

// 添加标注
drawLabel(1280, 430, '①', colors.envBorder, 24);
drawLabel(1080, 320, '②', colors.envBorder, 24);
drawLabel(600, 150, '③ 循环', colors.llmBorder, 20);
drawLabel(500, 400, '④ 完成', '#4CAF50', 20);

// 保存图片
const imagesDir = path.join(__dirname, 'images');
if (!fs.existsSync(imagesDir)) {
  fs.mkdirSync(imagesDir, { recursive: true });
}

const outputPath = path.join(imagesDir, 'react-workflow.png');
const buffer = canvas.toBuffer('image/png');
fs.writeFileSync(outputPath, buffer);

console.log('✅ 流程图已生成：', outputPath);


